include/identity_walker.h

```diff
@@ -44,36 +44,66 @@ enum WalkerStepResult
  * If clear_cache is set, the cache is cleared between calls to visit
  * otherwise, the cache persists
  *
+ * The user can optionally pass a pointer to a cache. If that pointer
+ * is non-null, it will be used in place of the internal cache.
+ *
  * Important Note: The term arguments should belong to the solver provided
  * to the identity walker, otherwise the behavior is undefined.
  */
 class IdentityWalker
 {
 public:
-  IdentityWalker(smt::SmtSolver & solver, bool clear_cache)
-    : solver_(solver), clear_cache_(clear_cache) {};
+ IdentityWalker(smt::SmtSolver & solver,
+                bool clear_cache,
+                smt::UnorderedTermMap * ext_cache = nullptr)
+     : solver_(solver), clear_cache_(clear_cache), ext_cache_(ext_cache){};
 
-  /** Visit a term and all its subterms in a pre- and post-order traversal
-   *  @param term the term to visit
-   *  @return the term after visiting (returns the value of cache[term]
-   *     -- if it has been cached and returns term otherwise)
-   */
-  smt::Term visit(smt::Term & term);
+ /** Visit a term and all its subterms in a pre- and post-order traversal
+  *  @param term the term to visit
+  *  @return the term after visiting (returns the value of cache[term]
+  *     -- if it has been cached and returns term otherwise)
+  */
+ smt::Term visit(smt::Term & term);
 
 protected:
-  /** Visit a single term.
-   *  Implement this method in a derived class to change the behavior
-   *  of the walker
-   *  @param term the term to visit
-   *  @return a WalkerStepResult to tell the visit method how to proceed
+ /** Visit a single term.
+  *  Implement this method in a derived class to change the behavior
+  *  of the walker
+  *  @param term the term to visit
+  *  @return a WalkerStepResult to tell the visit method how to proceed
+  */
+ virtual WalkerStepResult visit_term(smt::Term & term);
+
+ /** Check if key is in cache
+  *  @param key
+  *  @return true iff the key is in the cache
   */
-  virtual WalkerStepResult visit_term(smt::Term & term);
+ bool in_cache(const Term & key) const;
+
+ /** Query the cache. Automatically uses ext_cache_ if non-null.
+  *  @param key the term to check in the cache
+  *  @param out this term is set to the cache result if there's a hit
+  *  @return true iff there is a cache hit
+  */
+ bool query_cache(const Term & key, Term & out) const;
+
+ /** Populate the cache. Automatically uses ext_cache_ if non-null.
+  *  It will overwrite the existing mapping if the key is already in the cache
+  *  @param key the key term
+  *  @param val the value term
+  */
+ void populate_cache(const Term & key, const Term & val);
```

Comment:
can we change the name of this method? It sounds confusing to me. How about `save_in_cache`?

Reply:
Yep! Changed with e1fc7b40e57e7d5ac26c9ed053854ea21fc6a5c1